docs/DesignDocs/SwitchlessCalls.md

```diff
@@ -0,0 +1,84 @@
+Context-switchless Calls
+================
+
+## Motivation
+
+Context-switchless Calls are designed to reduce the cost of context switching between hosts and enclaves.
+In an enclave application, the host makes **ECALL**s into functions exposed by the enclaves it created. Likewise,
+the enclaves may make **OCALL**s into functions exposed by the host that created them. In either case, the
+execution has to be transitioned from an untrusted environment to a trusted environment, or vice versa. Since the
+transition is costly due to heavy security checks, it might be more performance advantageous to make the calls
+**context-switchless**: the caller delegates the function call to a worker thread in the other environment, which
+does the real job of calling the function and post the result to the caller. Both the calling thread and the
+worker thread never leave their respective execution contexts during the perceived function call.
+
+## Possible Usages
+
+In general, the good candidates for switchless calls are functions that are:
+1) short, thus the transition takes relatively high percentage of the overall execution time of the call; and
+2) called frequently, so the savings in transition time add up.
+
+If your enclave application has such functions and you are concerned with performance, consider making those
+functions context-switchless.
+
+## User Experience
+
+Firstly, The user has to identify which functions are good candidates of switchless calls. The identified ones
+need to be marked in the EDL file with a keyword `transition_using_threads`. For example, to mark function
+`host_increment_switchless` a target of switchless calls, it has to be declared as:
+
+```c
+void host_increment_switchless([in, out] int* m) transition_using_threads;
+```
+
+Secondly, while creating an enclave, the user has to explicitly configure it to enable switchless capability.
+An important setting in the configuration is how many worker threads are to be created for servicing the
+context-switchless calls. More worker threads typically means more competition for the CPU cores and more thread context switches, hurting the performance. On the other hand, less worker threads means means simultaneously issued switchless calls are less likely to be serviced quickly, if they got serviced at all.
```

Comment:
grammar: "less worker threads" -> "fewer worker threads"
(Also would be good to break this line into multiple lines in the source to make diffs easier to review)